### ISSUE TITLE
Identify actions needed to share ownership for EndToEnd with upstream teams

### ISSUE THREAD

**Issue by tmasternak** [Particular/EndToEnd#199]
## What this solves
In current setup it is not easy for upstream components to: update EndToEnd test to newest version and run the subset relevant to they on test infrastructure. This results in extended feedback loop from performance testing and potential risk of finding performance degradation problems to late. See https://github.com/Particular/PlatformDevelopment/issues/927 for more details

## Draft of PoA
Identify work needed to enable:
 - [ ] Easy package update for EndToEnd tests (maybe doco is enough)
 - [ ] Enable running subcategory of tests per component of interest
 - [ ] Enable easy execution of subcategory of tests in shared environment
 - [ ] Enable running test as part of build process of upstream components

NOTE: at least some of those could probably be solved with doco

/cc: @WojcikMike @Scooletz @SeanFeldman

**Comment by Scooletz** [comment]
Bumping up for sure. `Enable running subcategory of tests per component of interest` this could be combined with the last point by using a TeamCity parameter passed to the test runner. Each project would configure the one it is interested in.

**Comment by SeanFeldman** [comment]
Moving forward the "pain" is to be able to identify when a new feature or modification of an existing one is causing a performance regression. 

Ideally, I'd love to see this becoming a part of the automation we have in place, similar to unit tests, ATTs, and backwards compatibility tests. Challenge with this is to know what would indicate a performance regression. For that, a base line has to be established, and if my understanding is correct, we don't have one. E.g.: running on Azure <> running on EC2 <> running on static or dynamic.

I suspect this would involve many maintainer groups (such as transports and persistences) and whatever solution is going to be, needs to address concerns above plus concerns that maintainers already have.

**Comment by dvdstelt** [comment]
@tmasternak You mention EndToEnd tests, but which tests are you referring to? Because there are quite a lot of tests, physical solutions, etc. Performance, Wire Compat, etc.

How about this issue : https://github.com/Particular/PlatformDevelopment/issues/846
Apparently a fair amount of people involved, agreed on it, but there was never any action on it?

The fact that everything is still in a single repo, is probably the cause that I don't really understand what this issue is about.

**Comment by tmasternak** [comment]
@dvdstelt Let me try clarifying :) :
 * Particular/PlatformDevelopment#846 - this one is about splitting EndToEnd into separated repositories. The reason was that it is hard for current maintainer group to have knowledge about every type of tests that belong to EndToEnd
 * This issue and discussion in PlatDev Particular/PlatformDevelopment#927 was opened because I believe that currently the feedback loop between changing upstream component and getting any EndToEnd results in particular performance or backwards-compatibility tests is too big. One of the problems it causes in practice is that issues are/can be discovered when TF making changes no longer exists. The other is that it's sometimes hard to figure out if performance drop is actually a problem without proper context (see AtomicSendWithReceive case in ASB).

**Comment by Scooletz** [comment]
@SeanFeldman TeamCity supports gathering metrics and failing a build if a specific metrics is too high/low. That, combined with automated step of running these could be an indicator not only as a warning sign, but also, a real stopper, breaking any build that introduces this change.

**Comment by tmasternak** [comment]
> That, combined with automated step of running these could be an indicator not only as a warning sign, but also, a real stopper, breaking any build that introduces this change.

Agreed. But in ASB case we would make build green after proper investigation.

**Comment by dvdstelt** [comment]
> issues are/can be discovered when TF making changes no longer exists

This is where it got confusing for me. Do you mean that issues can be discovered (using EndToEnd projects) during the time the TF makes the changes and before it is disbanded?

In other words, you possibly want the EndToEnd projects to be updated, before NuGet (or MyGet) packages are released, and (for example) have it added to a `when is it done` policy? And then make it as easy as possible for these task forces to update EndToEnd repos themselves, likely in cooperation with the proper maintainer groups?

**Comment by tmasternak** [comment]
@dvdstelt I think that nuget releases are implementation details. 

I guess what I'm trying to say is that EndToEnd tests should be used as a tool by any TF doing changes to upstream components to verify the validity of their work. In other words it should be easy and possibly obligatory to run EndToEnd tests before calling issue done-done.

**Comment by ramonsmits** [comment]
I suggest to discuss this next week. We have a pretty good idea of what currently are pros and cons of the current solution. Besides informing maintainters, it's also important to come up with a different structure of the performance tests and maybe other tests so that we can get partial feedback. For example, when RabbitMQ or any of its (external) dependencies is updated to only update those statistics. Current solution is pretty dumb where for example all V5 tests are run again and again while there are not updates on any packages. Meaning smarter test execution, faster feedback and less costs.
